### PR TITLE
WIP: Certificate Service Extension

### PR DESCRIPTION
**Pull request by timuthy**
**What this PR does / why we need it**:
This PR adds an implementation of the Certificate Service Extension. It manages components like `Cert-Manager` and `Cert-Broker` which provide certificate services for Shoot clusters.

It's mainly comprised of 3 controllers:
- Lifecycle-Controller: manages Cert-Manager instance in the Seed.
- Certificate-Service-Controller: manages Cert-Broker instances in the Seed.
- RBAC-Controller: manages RBAC manifests in Shoot clusters.

**Which issue(s) this PR fixes**:
Fixes parts of https://github.com/gardener/gardener/issues/730

**Special notes for your reviewer**:
PR is WIP because tests are currently being worked on.

**Release note**:

```improvement operator
Gardener Extensions got a new extension controller which manages certificate service components.
```

### CODE REVIEW

File: controllers/extension-certificate-service/pkg/certservice/actuator.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+			if err := a.client.Delete(ctx, &obj); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+func createDNSProviderValuesForDomain(config configv1alpha1.DNSProviderConfig, shootDomain string) []interface{} {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/certservice/actuator.go]: How about putting these functions that generate chart values into an `internal` package and then test them there?

File: controllers/extension-certificate-service/pkg/certservice/add.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+	// SecretName is the name of the Secret which holds configuration information for the certificate-service.
+	SecretName = "certificate-service"
+
+	// watchBuilder determines which resources should be watched and the reconciler to be used.
+	watchBuilder controllerutil.WatcherBuilder
+)
+
+// Add adds a controller with the default Options to the given Controller Manager.
+func Add(mgr manager.Manager) error {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/certservice/add.go]: This should be named `AddToManager`

File: controllers/extension-certificate-service/pkg/certservice/add.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	return AddWithOptions(mgr, Options)
+}
+
+// AddWithOptions adds a controller with the given Options to the given manager.
+// The opts.Reconciler is being set with a newly instantiated actuator.
+func AddWithOptions(mgr manager.Manager, opts controller.Options) error {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/certservice/add.go]: This should be named `AddToManagerWithOptions`

File: controllers/extension-certificate-service/pkg/certservice/add.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	var (
+		cl                 = mgr.GetClient()
+		mapper             = typeMapperWithinNamespace(cl)
+		allNamespaceMapper = typeMapperForAllNamespaces(cl)
+	)
+
+	// Register Watch on Cluster resources.
+	watchBuilder.Register(func(ctrl controller.Controller) error {
+		if err := ctrl.Watch(
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/certservice/add.go]: Just directly return the `err` and don't do these checks.

File: controllers/extension-certificate-service/pkg/certservice/add.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		Name:              ControllerName,
+		Type:              Type,
+		ControllerOptions: opts,
+		WatchBuilder:      watchBuilder,
+	})
+}
+
+// typeMapperWithinNamespace returns a `ToRequestsFunc` that maps the incoming object
+// to the certificate service extension object in the same namespace.
+func typeMapperWithinNamespace(cl client.Client) handler.ToRequestsFunc {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/certservice/add.go]: I'd love to see a test for this.

File: controllers/extension-certificate-service/pkg/certservice/add.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+					Name:      resourceName,
+					Namespace: object.Meta.GetNamespace(),
+				},
+			}}
+	}
+}
+
+// typeMapperForAllNamespaces returns a `ToRequestsFunc` that maps the incoming object
+// to the certificate service extension objects across all namespaces.
+func typeMapperForAllNamespaces(cl client.Client) handler.ToRequestsFunc {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/certservice/add.go]: Please add a test.

File: controllers/extension-certificate-service/pkg/imagevector/imagevector.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+	"github.com/gardener/gardener-extensions/controllers/extension-certificate-service/pkg/utils"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+)
+
+// ImageVector is the image vector that contains al the needed images
+var ImageVector imagevector.ImageVector
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/imagevector/imagevector.go]: Please follow the same pattern as in the other controllers (not exposing the `ImageVector` var, reading the `ImageVector` via `packr`

File: controllers/extension-certificate-service/pkg/lifecycle/actuator.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+		applierOptions,
+	)
+}
+
+func (a *actuator) DeleteCertManager(ctx context.Context, secret *corev1.Secret) error {
+	certConfig, err := utils.DecodeCertConfig(secret)
+	if err != nil {
+		return fmt.Errorf("failed to decode certificate configuration: %v", err)
+	}
+	chartRenderer, err := chartrenderer.NewForConfig(a.config)
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/lifecycle/actuator.go]: You only need to initialize the `chartrenderer` once during `InjectConfig`.

File: controllers/extension-certificate-service/pkg/lifecycle/actuator.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+	}
+
+	var (
+		manifest   = release.Manifest()
+		decoder    = yaml.NewYAMLOrJSONDecoder(bytes.NewReader(manifest), 1024)
+		decodedObj map[string]interface{}
+	)
+
+	a.logger.Info("Component is being deleted", "component", "cert-manager")
+	for err = decoder.Decode(&decodedObj); err == nil; err = decoder.Decode(&decodedObj) {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/lifecycle/actuator.go]: Re-use potential if this is factored out.

File: controllers/extension-certificate-service/pkg/lifecycle/actuator.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+				return err
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func createCertServiceValues(certmanagementConfig *configv1alpha1.CertificateServiceConfigurationSpec) (map[string]interface{}, error) {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/lifecycle/actuator.go]: How about putting this into an `internal` package and writing tests for it?

File: controllers/extension-certificate-service/pkg/lifecycle/actuator.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+				},
+				"dns01": map[string]interface{}{
+					"providers": providers,
+				},
+			},
+		},
+	}, nil
+}
+
+func createDNSProviderValues(configs []configv1alpha1.DNSProviderConfig) ([]interface{}, error) {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/lifecycle/actuator.go]: Same as above.

File: controllers/extension-certificate-service/pkg/lifecycle/add.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+	// SecretName is the name of the Secret which holds configuration information for the certificate-service.
+	SecretName = "certificate-service"
+
+	// Sync determins the time the configuration of the certificate-service is reconciled.
+	Sync = 1 * time.Hour
+)
+
+// Add adds a Certificate Service Lifecycle controller to the given Controller Manager.
+func Add(mgr manager.Manager) error {
```
Comment (adracus) [inline controllers/extension-certificate-service/pkg/lifecycle/add.go]: Please rename to `AddToManager`.

File: pkg/controller/predicate.go [hunk 13]
```diff
@@ -129,3 +133,117 @@ func OrPredicate(predicates ...predicate.Predicate) predicate.Predicate {
 		},
 	}
 }
+
+// TypePredicate filters the incoming OperatingSystemConfigs for ones that have the same type
+// as the given type.
+func TypePredicate(typeName string) predicate.Predicate {
```
Comment (adracus) [inline pkg/controller/predicate.go]: I think those nice new predicates need to be tested :wink:

File: controllers/extension-certificate-service/cmd/app/app.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+	if err != nil {
+		controllercmd.LogErrAndExit(err, "Could not instantiate controller-manager")
+		return err
+	}
+
+	if err := gcontroller.AddToScheme(mgr.GetScheme()); err != nil {
+		controllercmd.LogErrAndExit(err, "Could not update manager scheme")
+	}
+
+	lifecycle.Options.MaxConcurrentReconciles = 1
```
[the remark this answers is not in the document]
Comment (timuthy) [reply controllers/extension-certificate-service/cmd/app/app.go]: The Lifecycle controller only reconciles one `Secret` which means that `MaxConcurrentReconciles = 1` is rather a static value. So no need to introduce a separate option and flag for it.
I can remove it here and add it directly to the instantiation of the [ControllerOptions](https://github.com/gardener/gardener-extensions/blob/49bc26908afd47a2c1767b29b9ee3efaf6522ea1/controllers/extension-certificate-service/pkg/lifecycle/add.go#L38).

**Comment by timuthy** [comment]
After some internal discussions we decided to restructure the controller, hence closing this in favor of #47.

**Pull request closed without merging**